src/support/threads.cpp

```diff
@@ -0,0 +1,201 @@
+        while (self->doWork() == ThreadWorkState::More) {}
+        self->doWork = nullptr;
+      } else if (self->done) {
+        DEBUG_PRINT(std::cerr << "[Thread] done\n");
+        return;
+      }
+    }
+    ThreadPool::get()->notifyThreadIsReady();
+    {
+      std::unique_lock<std::mutex> lock(self->mutex);
+      if (!self->done && !self->doWork) {
+        DEBUG_PRINT(std::cerr << "[Thread] thread waiting\n");
+        self->condition.wait(lock);
+      }
+    }
+  }
+}
+
+
+// ThreadPool
+
+void ThreadPool::initialize(size_t num) {
+  if (num == 1) return; // no multiple cores, don't create threads
+  DEBUG_PRINT(std::cerr << "[Pool] initialize()\n");
+  std::unique_lock<std::mutex> lock(mutex);
+  ready.store(threads.size()); // initial state before first resetThreadsAreReady()
+  resetThreadsAreReady();
+  for (size_t i = 0; i < num; i++) {
+    threads.emplace_back(std::unique_ptr<Thread>(new Thread()));
+  }
+  DEBUG_PRINT(std::cerr << "[Pool] initialize() waiting\n");
+  condition.wait(lock, [this]() { return areThreadsReady(); });
+  DEBUG_PRINT(std::cerr << "[Pool] initialize() is done\n");
+}
+
+ThreadPool* ThreadPool::get() {
+  if (!pool) {
+    assert(Thread::onMainThread());
+    size_t num = std::thread::hardware_concurrency();
+    if (num < 2) num = 1;
+    pool = std::unique_ptr<ThreadPool>(new ThreadPool());
+    pool->initialize(num);
+  }
+  return pool.get();
+}
+
+void ThreadPool::work(std::vector<std::function<ThreadWorkState ()>>& doWorkers) {
+  size_t num = threads.size();
+  // If no multiple cores, or on a side thread, do not use worker threads
+  if (num == 0 || !Thread::onMainThread()) {
+    // just run sequentially
+    DEBUG_PRINT(std::cerr << "[Pool] work() sequentially\n");
+    assert(doWorkers.size() > 0);
+    while (doWorkers[0]() == ThreadWorkState::More) {}
+    return;
+  }
+  // run in parallel on threads
+  // TODO: fancy work stealing
+  DEBUG_PRINT(std::cerr << "[Pool] work() on threads\n");
+  assert(doWorkers.size() == num);
+  assert(!running);
+  running = true;
+  std::unique_lock<std::mutex> lock(mutex);
+  resetThreadsAreReady();
+  for (size_t i = 0; i < num; i++) {
+    threads[i]->work(doWorkers[i]);
+  }
+  DEBUG_PRINT(std::cerr << "[Pool] main thread waiting\n");
+  condition.wait(lock, [this]() { return areThreadsReady(); });
+  DEBUG_PRINT(std::cerr << "[Pool] main thread waiting\n");
+  running = false;
+  DEBUG_PRINT(std::cerr << "[Pool] work() is done\n");
+}
+
+size_t ThreadPool::size() {
+  return threads.size();
+}
+
+bool ThreadPool::isRunning() {
+  return pool && pool->running;
+}
+
+void ThreadPool::notifyThreadIsReady() {
+  DEBUG_PRINT(std::cerr << "[Pool] notify thread is ready\n";)
+  std::lock_guard<std::mutex> lock(mutex);
+  ready.fetch_add(1);
+  condition.notify_one();
+}
+
+void ThreadPool::resetThreadsAreReady() {
+  DEBUG_PRINT(std::cerr << "[Pool] reset threads are ready\n";)
+  assert(ready.load() == threads.size());
+  ready.store(0);
```

Comment:
The assert and store are racy without the lock you have in `notifyThreadIsReady`. Some a huge deal since it's an assert, but still...